python/config_settings/transition.bzl

```
@@ -0,0 +1,200 @@
+"""The transition module contains the rule definitions to wrap py_binary and py_test and transition
+them to the desired target platform.
+"""
+
+load("@bazel_skylib//lib:dicts.bzl", "dicts")
+load("//python:defs.bzl", _py_binary = "py_binary", _py_test = "py_test")
+
+def _transition_python_version_impl(_, attr):
+    return {"//python/config_settings:python_version": str(attr.python_version)}
+
+_transition_python_version = transition(
+    implementation = _transition_python_version_impl,
+    inputs = [],
+    outputs = ["//python/config_settings:python_version"],
+)
+
+def _transition_py_impl(ctx):
+    target = ctx.attr.target[0]
+    executable = ctx.actions.declare_file(ctx.attr.name + (".exe" if ctx.attr.is_windows else ""))
+    ctx.actions.symlink(
+        is_executable = True,
+        output = executable,
+        target_file = target[DefaultInfo].files_to_run.executable,
+    )
+    zipfile_symlink = None
+    if ctx.attr.is_windows:
+        zipfile = None
+        expected_target_path = target[DefaultInfo].files_to_run.executable.short_path[:-4] + ".zip"
+        for file in target[DefaultInfo].default_runfiles.files.to_list():
+            if file.short_path == expected_target_path:
+                zipfile = file
+        zipfile_symlink = ctx.actions.declare_file(ctx.attr.name + ".zip")
+        ctx.actions.symlink(
+            is_executable = True,
+            output = zipfile_symlink,
+            target_file = zipfile,
+        )
+    env = {}
+    for k, v in ctx.attr.env.items():
+        env[k] = ctx.expand_location(v)
+    providers = [
+        DefaultInfo(
+            executable = executable,
+            files = depset([zipfile_symlink] if zipfile_symlink else [], transitive = [target[DefaultInfo].files]),
+            runfiles = ctx.runfiles([zipfile_symlink] if zipfile_symlink else []).merge(target[DefaultInfo].default_runfiles),
+        ),
+        target[PyInfo],
+        target[PyRuntimeInfo],
+        target[InstrumentedFilesInfo],
```

To make sure the coverage instrumentation filters match the wrapped binary, specify the target attribute as one of the dependency attributes. The other attributes with the common names (srcs, deps, etc) should be automatically handled.
```suggestion
        # Ensure that the binary we're wrapping is included in code coverage.
        coverage_common.instrumented_files_info(
            ctx,
            dependency_attributes = ["target"],
        ),      
```